### DISCUSSION TITLE
Client won't check in with API

### DISCUSSION THREAD

**Discussion by tm6907** [cmu-sei/GHOSTS#280] [Q&A] [answered]
Hi, really cool project by the way. I've been struggling to get it all working and I'm not sure why (not the most experienced Linux or Docker user). The client works fine alone, browsing websites on Firefox etc.

I have a Rocky 9 Linux VM running the Ghosts - API with docker installed and running the three containers. I then have a separate win10 VM with the Ghosts - Client installed and running.

The application.json file should be configured correctly, but the client never checks into the API. The localhost:5000/api/home site never updates from 0 machines. Depends if the url should be the IP of the host VM itself or the IP of the API-container (I've tried both). 

From the win10 client I can ping the other VM (Linux host), the bridge IP (172.17.0.1) and docker0 (172.18.0.1). But can't ping any of the containers i.e. the API container 172.18.0.3.

There's no id.json file in the /instance folder but there are entries in the app, clientupdates and clientupdates_not_posted logs. The main error repeating is "Skipping check for ID from server..."

**Comment by sei-dupdyke** [comment]
Hi, thanks for checking out the project.

1. On the client - the URL that you are using in `config/application.json` for the API — can you hit that same URL in the browser on the client machine?

2. On the server — do you see anything in the logs for the API? try `docker logs ghosts-api` and see if there are any errors.

LMK and we'll get you squared away.

**Reply by tm6907** [reply]
Ok managed to get a Linux client and the Win10 client to check in, turns out I needed to add an exception to the proxy settings for the host VM. Next step is to make them do some NPC stuff. 

Could I ask, is there a way to de-register a client from the host server? I want to test what security I can have enabled or not for the client to still connect. E.g. I disabled the windows firewall and want to test with it on. Cheers

**Comment by sei-dupdyke** [comment]
There is no way to do this currently from the API, but you could do it manually via:

On the client:
- find the guid for this client in ./instance/id.json
- delete the ./instance folder

On the server:
```bash
docker exec -it ghosts-postgres /bin/bash
psql -U ghosts
delete * from machines where id = 'the guid of the machine you want to delete' CASCADE;
```

**Reply by tm6907** [reply]
Ah thanks, I'll give that a whirl. 

Another issue I've discovered is the geckodriver isn't working on the linux client. 

![image](https://github.com/cmu-sei/GHOSTS/assets/162140616/fde705ca-37dc-45b2-8cfd-6a490a3c98e4)

I've double checked the permissions and tried running the .exe from the file and cmd but it won't execute. Plus I've altered the environment variable to point to where the .exe is installed with: 
"export PATH=$PATH:/home/user/ghosts/ghost-client-linux-v7.0.0" 

Any idea whats going wrong? Thanks

**Reply by sei-dupdyke** [reply]
You have to match the version of geckodriver that works with the version of FF that you are using: 

https://github.com/mozilla/geckodriver/releases

**Reply by tm6907** [reply]
All sorted now, I ended up installing the wrong geckodriver for the architecture on the VM. Both clients are browsing away now, next is to make them do other NPC stuff :)

**Reply by tm6907** [reply]
Sorry, getting a syntax error running the last command you mentioned. 

"""
delete from machines where id = 'the guid of the machine you want to delete' CASCADE;
""""

Syntax error near the *

**Reply by sei-dupdyke** [reply]
Whoops, no "*"! Sorry.

**Reply by tm6907** [reply]
Ah brill,

Edit: sorry I don't know how to work this
<img width="379" alt="image" src="https://github.com/cmu-sei/GHOSTS/assets/162140616/60305cd3-cc3c-4ca4-aaa6-08f00ed2f514">

**Reply by tm6907** [reply]
Ah nevermind sorted it, I was just being daft. 

Thanks for your help :)

**Reply by tm6907** [reply]
Hi raising this issue again, trying to restart a client because Im not seeing any tcp traffic between it and the host. However, I can't seem to remove it from the postgres database like I could before. 

<img width="876" alt="image" src="https://github.com/cmu-sei/GHOSTS/assets/162140616/580d828f-5351-4af8-bb79-51fa39fe0922">
<img width="668" alt="image" src="https://github.com/cmu-sei/GHOSTS/assets/162140616/858f1401-c5d7-4ba5-b320-a7613679d95b">

**Reply by sei-dupdyke** [reply, marked as answer]
Hi — try 'guid' instead of "guid", postgres does not like double quotes. As an example:

```sql
select * from machines where id = '3fff58d0-caa0-4569-9159-1c8f2baf866c';
```
